[Merged by Bors] - feat(topology/connected): Connectedness of unions of sets

AntoineChambert-Loir opened the pull request:
* Add multiple results about when unions of sets are (pre)connected. In particular, the union of connected sets indexed by `ℕ` such that each set intersects the next is connected.
* Remove some `set.` prefixes in the file
* There are two minor fixes in other files, presumably caused by the fact that they now import `order.succ_pred`
* Co-authored by Floris van Doorn fpvdoorn@gmail.com

---

- [x] depends on: #11518

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Commit:
feat(src/topology/connected.lean): add a theorem on connected unions

Add a theorem that unions of preconnected sets indexed by integers
such that each of them meets the next one is preconnected

In a separate file (src/data/set/increasing_union.lean),
add a series of possibly lemma on increasing unions
It is possible that these lemmas have a neater proof.

jcommelin commented on src/data/set/increasing_union.lean:
```diff
@@ -0,0 +1,104 @@
+
```
In VScode, you can type `copy<Tab>` to get a template for the copyright header.

AntoineChambert-Loir replied:
I just did it!

jcommelin commented on src/data/set/increasing_union.lean:
```diff
@@ -0,0 +1,104 @@
+
+/- Given a sequence of sets s : ℕ → set α, 
```
```suggestion
/-! # Please add some title here

Given a sequence of sets s : ℕ → set α, 
```

AntoineChambert-Loir replied:
I added a title.

Commit:
style(src/data/set/increasing_union.lean): add copyright and title

eric-wieser commented on src/data/set/increasing_union.lean:
```diff
@@ -0,0 +1,112 @@
+-/
+
+/-
+# Increasing unions of sets indexed by integers
```
```suggestion
# Increasing unions of sets indexed by naturals
```

eric-wieser commented on src/data/set/increasing_union.lean:
```diff
@@ -0,0 +1,112 @@
+
+Given a sequence of sets s : ℕ → set α,
+we construct its increasing Union by
+λ n:ℕ, Union (set restrict s { j : ℕ | j ≤ n}) -/
```
Are you sure you don't want
```suggestion
λ n:ℕ, Union (subtype.restrict s (≤ n)) -/
```

eric-wieser replied:
This also looks a lot like `partial_sups`, cc @YaelDillies

urkud replied:
And `accumulate`.

YaelDillies replied:
Yeah it is `partial_sups`. I'm sorry you wrote it all, Antoine, because this is duplicate 😕

AntoineChambert-Loir replied:
I was almost certain something like that would exist, and I am truly glad that it already exists. 
I'll try to understand how it works tomorrow and to adjust the other file accordingly.

AntoineChambert-Loir replied:
@YaelDillies , OK, it was nice and easy to use `partial_sups`.
At first, I thought that the assertion that `partial_sups s` is monotone was missing, but it seems to be included in the galois connection…

YaelDillies replied:
`partial_sups` is *defined* as a monotone function. That's what the little `m` to the right of the arrow stands for. You can look at the API for `preorder_hom` for how to use this.

AntoineChambert-Loir replied:
Unfortunately, I can't read that little symbol on the right of the arrow, it is printed as a box in both vim and VisualStudio, all I know is that it has Unicode code 8344…

jcommelin replied:
@AntoineChambert-Loir In that case I think you are in dire need of a font with more unicode support :wink:

AntoineChambert-Loir replied:
That I could guess, but I think I use the standard VisualStudioCode (on Mac OS)…

urkud commented on src/topology/connected.lean:
```diff
@@ -132,6 +134,167 @@ begin
+
+theorem is_preconnected.Union_of_directed   { s : ι → set α }  
+  (H : ∀ n: ι, is_preconnected (s n))
+  (K : ∀ m n: ι, m ≤ n → s m ⊆ s n) : 
```
You require monotonicity, not `directed` here.

AntoineChambert-Loir replied:
Yes. I changed the terminology.

Commit:
style: use partial_sups and delete useless additions of first commit

Commit:
style: suppress invocation of deleted file

Commit:
style: add the necessary import

Commit:
style: remove noisy #lint

eric-wieser commented on src/topology/connected.lean:
```diff
@@ -132,6 +133,165 @@ begin
+is preconnected
+-/
+variable ι : Type _
+variables [ linear_order ι ]
```
the mathlib style would be:
```suggestion
variables {ι : Type*} [linear_order ι]
```

eric-wieser commented on src/topology/connected.lean:
```diff
@@ -132,6 +133,165 @@ begin
+variable ι : Type _
+variables [ linear_order ι ]
+
+theorem is_preconnected.Union_of_monotone   { s : ι → set α }  (K : monotone s)
```
mathlib style is single spaces, no spaces within binder brackets:
```suggestion
theorem is_preconnected.Union_of_monotone {s : ι → set α} (K : monotone s)
```
this applies throughout.

eric-wieser commented on src/topology/connected.lean:
```diff
@@ -1008,3 +1168,4 @@ lemma continuous.connected_components_map_continuous {β : Type*} [topological_s
 continuous.connected_components_lift_continuous (continuous_quotient_mk.comp h)
 
 end connected_component_setoid
+
```
```suggestion
```

eric-wieser commented on src/topology/connected.lean:
```diff
@@ -132,6 +133,165 @@ begin
+      apply has_subset.subset.trans (subset_Union s n) Huv,
+      apply nonempty_of_mem  (mem_inter Kan Kua.right),
+      apply nonempty_of_mem (mem_inter Kbn Kvb.right),
+  end,
```
```suggestion
  have Hnuv : ((s n) ∩ (u ∩ v)).nonempty,
  { apply (H n), apply hu, apply hv,
    apply has_subset.subset.trans (subset_Union s n) Huv,
    apply nonempty_of_mem  (mem_inter Kan Kua.right),
    apply nonempty_of_mem (mem_inter Kbn Kvb.right), },
```
mathlib style is to use `, { sorry }` not `:= begin sorry end` for `have`s inside a tactic block

eric-wieser commented on src/topology/connected.lean:
```diff
@@ -132,6 +133,165 @@ begin
+theorem is_preconnected.filtered_sUnion_of { S : set (set α) }
+  (K : ∀ s t, s ∈ S → t ∈ S → s ⊆ t ∨ t ⊆ s)
+  (H : ∀ s, s ∈  S →  is_preconnected s ) :
+  is_preconnected ( sUnion S ) :=
```
style: too many spaces again
```suggestion
theorem is_preconnected.filtered_sUnion_of {S : set (set α)}
  (K : ∀ s t, s ∈ S → t ∈ S → s ⊆ t ∨ t ⊆ s)
  (H : ∀ s, s ∈ S → is_preconnected s) :
  is_preconnected (sUnion S) :=
```

eric-wieser reviewed (commented):
Some initial style comments

AntoineChambert-Loir committed:
Update src/topology/connected.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

AntoineChambert-Loir committed:
Update src/topology/connected.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Commit:
style: remove double spaces and begin/end in have

Commit:
Merge branch 'ad-connectedness' of github.com:leanprover-community/mathlib into ad-connectedness

urkud commented on src/topology/connected.lean:
```diff
@@ -132,6 +133,150 @@ begin
 end
 
+/--
+The union Union s of an increasing mapping s of preconnected sets
```
Is there a typo here? If not, how should I interpret "The union Union s"?

YaelDillies replied:
You might also want to format using ``:
```suggestion
The union `Union s` of an increasing mapping `s` of preconnected sets
```
but `Union s` should use the `⋃` notation regardless.

AntoineChambert-Loir replied:
What do you mean by using the ``⋃``notation?

PatrickMassot committed:
Use some tricks of the trade

eric-wieser commented on src/topology/connected.lean:
```diff
@@ -134,92 +134,69 @@ begin
+We could also use
+def set.is_filtered' {α : Type*} (S : set (set α)) := zorn.chain (⊆) S
+which would be equivalent but only asks:
+∀ s t, s ∈ S → t ∈ S → → s ≠ t → s ⊆ t ∨ t ⊆ s
```
Typo? `→ →` looks wrong.

Commit:
style(topology/connected.lean): implement Patrick Massot's suggestions and do minor rewrites

Commit:
style(topology/connected.lean): split two lines

PatrickMassot committed:
Compress a bit

fpvandoorn commented on src/topology/connected.lean:
```diff
@@ -132,6 +133,73 @@ begin
+  rw union_comm at Huv,
+  rw inter_comm u v,
+  exact this t s b a v u hv hu Huv hbv hbt hau has,
+end
```
```suggestion
/-- The directed sUnion of a set S of preconnected subsets is preconnected. -/
theorem is_preconnected.sUnion_directed {S : set (set α)}
  (K : directed_on (⊆) S)
  (H : ∀ s ∈ S, is_preconnected s) : is_preconnected (⋃₀ S) :=
begin
  rintros u v hu hv Huv ⟨a, ⟨s, hsS, has⟩, hau⟩ ⟨b, ⟨t, htS, hbt⟩, hbv⟩,
  obtain ⟨r, hrS, hsr, htr⟩ : ∃ r ∈ S, s ⊆ r ∧ t ⊆ r := K s hsS t htS,
  have Hnuv : (r ∩ (u ∩ v)).nonempty,
  from H _ hrS u v hu hv ((subset_sUnion_of_mem hrS).trans Huv) ⟨a, hsr has, hau⟩ ⟨b, htr hbt, hbv⟩,
  have Kruv : r ∩ (u ∩ v) ⊆ ⋃₀ S ∩ (u ∩ v),
  from inter_subset_inter_left _ (subset_sUnion_of_mem hrS),
  exact Hnuv.mono Kruv
end
```
This statement can be generalized by only requiring `S` to be directed, which makes the proof also significantly simpler.

fpvandoorn commented on src/topology/connected.lean:
```diff
@@ -132,6 +133,73 @@ begin
+theorem is_preconnected.union_of_chain { s : ℕ → set α }
+  (H : ∀ n: ℕ, is_preconnected (s n))
+  (K : ∀ n: ℕ, (s n ∩ s(n.succ)).nonempty) :
+  is_preconnected (⋃ n, s n) :=
```
I feel like this lemma should be generalized. With lemmas like this, it's really tricky to apply this lemma if you want a different indexing type, like Patrick's proof of the finite case shows. Therefore, you want to try to generalize a lemma like this to arbitrary indexing types.

It is a little tricky to state this lemma for an arbitrary indexing type, but I think the following is pretty nice:
```lean
theorem is_preconnected.Union_of_trans_gen {ι} {s : ι → set α}
  (H : ∀ i, is_preconnected (s i))
  (K : ∀ i j, relation.trans_gen (λ i j : ι, (s i ∩ s j).nonempty) i j) :
  is_preconnected (⋃ n, s n)
```
In words, condition `K` states: the transitive closure of the relation `(λ i j : ι, (s i ∩ s j).nonempty)` is the full relation.

Together with a lemma like the following (which I think doesn't exist yet):
```lean
lemma relation.trans_gen_nat (r : ℕ → ℕ → Prop) {n m : ℕ} 
  (h : ∀ i ∈ Ico (min n m) (max n m), r i i.succ ∧ r i.succ i) : relation.trans_gen r n m := 
sorry
```
we can easily prove your original lemma
```lean
theorem is_preconnected.Union_nat_of_chain { s : ℕ → set α }
  (H : ∀ n: ℕ, is_preconnected (s n))
  (K : ∀ n: ℕ, (s n ∩ s(n.succ)).nonempty) :
  is_preconnected (⋃ n, s n) :=
is_preconnected.Union_of_chain' H $ λ i j, 
  relation.trans_gen_nat _ (λ i _, ⟨K i, by { rw inter_comm, exact K i }⟩)
```
I'm sure that with this formulation, something like Patrick's case is also easy to prove.

I'm happy to help with proving this more general case, if you want.

fpvandoorn replied:
To prove this more general version, we probably need a Zorn-like argument.

AntoineChambert-Loir replied:
@fpvandoorn (I had answered your remark by email, but it is better to write it here.)

I like your formulation in so that it exactly says that the union a family of preconnected sets indexed by the vertices of a connected graph is preconnected if the sets indexed by any two adjacent vertices meet.

Together with a list of relevant graphs, this would be fine.

By the way, the case (in mathlib) of a union of connected sets with a given common point is a particular case of this (the graph contains a star), and the other interesting graphs to consider are intervals of Z.

AntoineChambert-Loir replied:
Also, it might have been easier to prove this theorem using the criterion that any continuous map to the {0,1} -space is constant.

AntoineChambert-Loir replied:
@fpvandoorn, I could prove the theorem along the lines you indicate, but I have made a detour through `refl_trans_gen`, whose induction principle was clearly stated in mathlib ('logic/relation.lean'), and this forced me to treat the case of connected sets.
The difference is of very little importance, but one never knows.
(One needs to assume that \iota is nonempty.)

(I still need to add the lemma that the transitive closure of the succ relation on any interval of Z is the full relation.)

AntoineChambert-Loir replied:
@fpvandoorn, I could prove the theorem along the lines you indicate, but I have made a detour through `refl_trans_gen`, whose induction principle was clearly stated in mathlib ('logic/relation.lean'), and this forced me to treat the case of connected sets.
The difference is of very little importance, but one never knows.
(One needs to assume that \iota is nonempty.)

(I still need to add the lemma that the transitive closure of the succ relation on any interval of Z is the full relation.)

fpvandoorn replied:
Good! Can you push your work to this branch so that we can review it?

It's probably slightly nicer to do it with `trans_gen` instead of `refl_trans_gen`, so that you don't have to assume non-emptyness. You can do induction on either predicate by using the tactic `induction h` if `h` has type `trans_gen x y` (or `refl_trans_gen x y`). The two induction principles `refl_trans_gen.head_induction_on` and `refl_trans_gen.trans_induction_on` are alternative induction principles (we could easily prove similar results for `trans_gen` if you need them). 
But this is minor, and doesn't necessary have to change for this PR.

AntoineChambert-Loir replied:
I can try to do this, but I don't know how to do it. Can I  just add a  new bunch of functions below the previous ones, with new names, and commit/push the whole stuff?

fpvandoorn replied:
Yes, that sounds good. I'll clean it up a bit!

AntoineChambert-Loir replied:
The following lemma 
```

lemma succ_gen_of_int_interval (ι:set ℤ) [hι: ord_connected ι]: 
  ∀(i j:ℤ), i ∈ ι → j ∈ ι → relation.refl_trans_gen
     (λ i j:ℤ, i ∈ ι ∧ j ∈ ι ∧ (j = i+1 ∨ i = j+ 1)) i j
  :=
begin
  intros m n hm, 
  refine int.induction_on' n m _ _ _ ,
  -- case: m m
  { intro _, exact relation.refl_trans_gen.refl, }, 
  -- forward induction : m k → m (k+1)
  { intros k hmk IHk hk',
    apply (relation.refl_trans_gen.cases_tail_iff _ m (k+1)).2,
    have hk : k ∈ ι,
      { apply set.ord_connected.out hι hm hk',
        simp only [zero_le_one, and_true, le_add_iff_nonneg_right, mem_Icc, hmk]},
    apply or.intro_right _ , use k, split,
    exact IHk hk,
    apply and.intro hk, apply and.intro hk', 
    simp only [true_or,eq_self_iff_true] }, 
  -- backward inductin : m k → m (k-1)
  { intros k hmk IHk hk',
    have hk : k ∈ ι,
      { apply set.ord_connected.out hι  hk' hm,
      simp only [hmk, zero_le_one, sub_le_self_iff, and_self, mem_Icc]},
    apply (relation.refl_trans_gen.cases_tail_iff _ m (k-1)).2,
    apply or.intro_right _ , use k, split,
    exact IHk hk,
    apply and.intro hk, apply and.intro hk', 
    simp only [sub_add_cancel, eq_self_iff_true, or_true]}  
end
```
proves that int intervals will satisfy the required property.
However, I can't use it because I cannot write something as :
```
theorem is_connected.Union_nat_of_int_interval 
  { ι : set ℤ} {hι₀:nonempty ι} {hι: set.ord_connected ι} 
  { s : ι → set α }
  (H : ∀ i: ι, is_connected (s i))
  (K : ∀ i: ℤ, i ∈ ι → i.succ ∈ ι → (s i ∩ s i.succ).nonempty) :
  is_connected (⋃ i, s i) :=
```

there is a type mismath at `s i` on line 5. A problem of coercion I don't know how to code in Lean.

fpvandoorn replied:
I modified the proofs to work for `preconnected`. 
I reformulated the `Union` lemma a little bit so that the graph is defined by the non-emptyness conditions.
I also reformulated (generalized) the lemmas about `nat`-relations, and move them to a separate file (a new one, since there was no file that already existed that imported the right files where it was appropriate).
I proved similar lemmas for `int`-relations. The formulation I added should be general enough to reason about any `trans_gen` on integers.

To formulate the right lemma on the integers, you probably want to make `s : \Z → set α` but work with `⋃ i \in \iota, s i`.
The most convenient way to do this is to also prove `is_preconnected.Union_of_trans_gen` for a union of the form `⋃ i \in \iota, s i` (we call these forms of unions `bUnion` for bounded union in the library).

fpvandoorn replied:
I added the corresponding induction principles for `trans_gen` in #10331, but didn't end up using them for this PR.

AntoineChambert-Loir replied:
Do you want me to go on or will you do it yourself?

fpvandoorn replied:
If you're happy to try it, please do! `is_preconnected.bUnion_of_trans_gen` should come before `is_preconnected.Union_of_trans_gen` with appropriate extra conditions (everything is restricted to a subset of `\iota`), and then `is_preconnected.Union_of_trans_gen` should be an easy consequence.

AntoineChambert-Loir replied:
Good. I'll probably do it tomorrow. Thanks for the help !

Commit:
feat(topology/connected.lean): add a theorem on connectedness of union

Prove that the Union of connected sets indexed by the vertices of a
graph is connected, when the graph is connected and any two sets
indexed by neighboring vertices meet.

Apply this to a new proof that the union of sets indexed by `nat`
is connected when a set and the next meet.

TODO: - apply this to sets indexed by an interval in `int`.

Commit:
feat(connected.lean): generalize the equiv relation stuff to intervals

fpvandoorn committed:
prove properties for preconnectedness

fpvandoorn committed:
add int relation

fpvandoorn committed:
a little bit of cleanup

fpvandoorn committed:
style linter

fpvandoorn committed:
Merge remote-tracking branch 'origin/master' into ad-connectedness

fpvandoorn committed:
style linter

fpvandoorn committed:
fix indent

YaelDillies commented on src/data/int/order.lean:
```diff
@@ -89,4 +89,17 @@ begin
 end
 
+/-- Induction principle on `ℤ` for all integers above a given integer `m`. -/
+@[elab_as_eliminator] lemma le_induction {P : ℤ → Prop} {m : ℤ}
```
This could be a recurrence principle. While you're at it, you should provide the dual `ge_rec`.
```suggestion
@[elab_as_eliminator] def le_rec {P : ℤ → Sort*} {m : ℤ}
```

fpvandoorn replied:
The current proof works for `Prop` only, so I prefer to not do this in this PR.

AntoineChambert-Loir replied:
Mathlib's #data.nat.basic has `nat.le_induction` that works for `Prop`, and `nat.le_rec_on` that works in general. Why not the same for `int` ?

AntoineChambert-Loir replied:
(I am not sure I understand the issue, but using `obtain` or other `dcases` eliminators seems to be restricted for `Prop`-valued functions. I hoped eliminating a `exists.unique` would help, but it does not seem to do.

AntoineChambert-Loir replied:
(Now, I guess that there is a computability issue, but it should not be serious in the case of `int` where equality is decidable.)

fpvandoorn replied:
It is definitely possible to do define `int.le_rec[_on]` as well, but it takes some care if you want it to compute in a good way. For `nat` this is easier, since the inequality is defined directly as an inductive type (see `nat.less_than_or_equal`).

Commit:
connectedness of bUnion and the case of integers

fpvandoorn committed:
fix

fpvandoorn committed:
some simplifications

fpvandoorn committed:
prepare mathlib bump

fpvandoorn committed:
Merge remote-tracking branch 'origin/master' into ad-connectedness

fpvandoorn committed:
fix

fpvandoorn committed:
Merge remote-tracking branch 'origin/master' into ad-connectedness

YaelDillies commented on src/data/nat/relation.lean:
```diff
@@ -0,0 +1,55 @@
+
+/-- `(n, m)` is in the reflexive-transitive closure of `~` if `i ~ i+1` and `i+1 ~ i`
+  for all `i` between `n` and `m`. -/
+lemma refl_trans_gen_nat (r : ℕ → ℕ → Prop) {n m : ℕ}
```
You should have `nat.refl_trans_gen_of_le` the one-sided version. Then this one is a one-liner. Same for `int`.

YaelDillies commented on src/topology/connected.lean:
```diff
@@ -4,6 +4,10 @@ Released under Apache 2.0 license as described in the file LICENSE.
+import order.partial_sups
+import logic.relation
+import data.nat.relation
+import data.int.relation
```
```suggestion
import data.int.relation
import data.nat.relation
import order.partial_sups
import topology.subset_properties
```

YaelDillies commented on src/logic/relation.lean:
```diff
@@ -390,6 +390,10 @@ lemma trans_gen.closed {p : α → α → Prop} :
   (∀ a b, r a b → trans_gen p a b) → trans_gen r a b → trans_gen p a b :=
 trans_gen.lift' id
 
+lemma trans_gen.mono {p : α → α → Prop} :
```
Can you provide `refl_gen.mono` as well?

YaelDillies commented on src/data/nat/relation.lean:
```diff
@@ -0,0 +1,55 @@
+
+/-- `(n, m)` is in the transitive closure of a relation `~` for `n ≠ m` if `i ~ i+1` and `i+1 ~ i`
+  for all `i` between `n` and `m`. -/
+lemma trans_gen_nat_of_ne (r : ℕ → ℕ → Prop) {n m : ℕ}
```
This feels like a better name.
```suggestion
lemma nat.trans_gen_of_ne (r : ℕ → ℕ → Prop) {n m : ℕ}
```

YaelDillies commented on src/data/int/relation.lean:
```diff
@@ -0,0 +1,60 @@
+
+/-- `(n, m)` is in the reflexive-transitive closure of `~` if `i ~ i+1` and `i+1 ~ i`
+  for all `i` between `n` and `m`. -/
+lemma refl_trans_gen_int (r : ℤ → ℤ → Prop) {n m : ℤ}
```
All those lemmas should be about `[linear_order α] [succ_order α] [is_succ_archimedean α]`. See for example `succ.rec_linear`.

YaelDillies reviewed (commented):
I'm not sure these new relation lemmas deserve two new files, especially considering that they are a continuation of `order.succ_pred`.

fpvandoorn committed:
Merge remote-tracking branch 'origin/master' into ad-connectedness

fpvandoorn committed:
some preliminary fixes, probably still broken

fpvandoorn committed:
fix

fpvandoorn committed:
generalize Z

fpvandoorn committed:
sort imports

fpvandoorn committed:
fix

fpvandoorn committed:
I don't know why this simplified

fpvandoorn committed:
Merge remote-tracking branch 'origin/master' into ad-connectedness

fpvandoorn committed:
rename some results and generalize nat to succ_order

jcommelin reviewed (approved):
Thanks :tada: It took a while, but this looks really great now!

bors merge

YaelDillies commented on src/topology/connected.lean:
```diff
@@ -215,28 +215,32 @@ theorem is_connected.Union_of_refl_trans_gen {ι : Type*} [nonempty ι] {s : ι
-theorem is_connected.Union_nat_of_chain {s : ℕ → set α}
-  (H : ∀ n : ℕ, is_connected (s n))
-  (K : ∀ n : ℕ, (s n ∩ s n.succ).nonempty) :
+/-- The Union of connected sets indexed by a type with an archimedian successor (like `ℕ` or `ℤ`)
```
```suggestion
/-- The Union of connected sets indexed by a type with an archimedean successor (like `ℕ` or `ℤ`)
```

YaelDillies commented on src/topology/connected.lean:
```diff
@@ -215,28 +215,32 @@ theorem is_connected.Union_of_refl_trans_gen {ι : Type*} [nonempty ι] {s : ι
+
+/-- The Union of connected sets indexed by a type with an archimedian successor (like `ℕ` or `ℤ`)
+  such that any two neighboring sets meet is preconnected. -/
+theorem is_preconnected.Union_of_chain {s : β → set α}
```
I would get rid of dot notation here and below because you can't actually use it.

fpvandoorn replied:
I agree you cannot actually use it (though I kinda hope we can change the way projection notation works in Lean 4 so that `H.Union_of_chain K` works in this case).
However, I'm not sure if that means we should change the `.` to a `_`: it might be more predictable to guess the name if we consistently keep using `.` even when projection notation doesn't apply.

YaelDillies replied:
At least, there's precedent for not using it when the condition is in a forall. For example, `star_convex_Union`.

fpvandoorn replied:
I believe there is a bunch of precedent for both options. We have not decided on a rule here. (`measurable_set.pi` to pick one)
